File: EgoCapture/Models/IMUSample.swift
import Foundation

/// Represents a single synchronized accelerometer + gyroscope sample from CoreMotion.
/// Coordinates follow Apple's device coordinate system (right-hand rule, Z out of screen).
struct IMUSample: Codable {
    /// Epoch timestamp in milliseconds (wall clock — for logging only)
    let timestampEpochMs: Double
    /// Milliseconds elapsed since recording start
    let relativeMs: Double
    /// Monotonic timestamp in nanoseconds (mach_absolute_time based)
    let timestampNs: UInt64
    /// Clock source identifier
    let clock: String // "mach_absolute_time"
    /// Accelerometer reading in G's (gravity-included via CMDeviceMotion)
    let accelerometer: XYZ
    /// Gyroscope reading in radians/second
    let gyroscope: XYZ
    
    struct XYZ: Codable {
        let x: Double
        let y: Double
        let z: Double
    }
}

File: android/app/src/main/java/com/kgeneye/eye/capture/ImuSample.kt
package com.kgeneye.eye.capture

/**
 * Synchronized accelerometer + gyroscope sample.
 *
 * Axis convention mirrors the device reference frame reported by
 * [android.hardware.SensorManager] (right-hand rule, +Z out of the screen).
 * Accelerometer values are stored in **G's** (1 G ≈ 9.80665 m/s²) so the
 * payload is comparable to the iOS `IMUSample` captured via CoreMotion.
 */
data class ImuSample(
    /** Epoch wall-clock in milliseconds (logging/debug only). */
    val timestampEpochMs: Double,
    /** Milliseconds elapsed since recording start. */
    val relativeMs: Double,
    /** Monotonic nanoseconds ([android.os.SystemClock.elapsedRealtimeNanos]). */
    val timestampNs: Long,
    /** Clock source identifier. */
    val clock: String,
    val accelerometer: XYZ,
    val gyroscope: XYZ,
) {
    data class XYZ(val x: Double, val y: Double, val z: Double)
}
